time.go

```diff
@@ -0,0 +1,96 @@
+package null
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+)
+
+// Time is a nullable time.Time. It supports SQL and JSON serialization.
+// It will marshal to null if null.
+type Time struct {
+	Time  time.Time
+	Valid bool
+}
+
+// Scan implements the Scanner interface.
+func (t *Time) Scan(value interface{}) error {
+	t.Time, t.Valid = value.(time.Time)
+	return nil
+}
+
+// Value implements the driver Valuer interface.
+func (t Time) Value() (driver.Value, error) {
+	if !t.Valid {
+		return nil, nil
+	}
+	return t.Time, nil
+}
+
+// NewTime creates a new Time.
+func NewTime(t time.Time, valid bool) Time {
+	return Time{
+		Time:  t,
+		Valid: valid,
+	}
+}
+
+// TimeFrom creates a new Time that will always be valid.
+func TimeFrom(t time.Time) Time {
+	return NewTime(t, true)
+}
+
+// TimeFromPtr creates a new Time that will be null if t is nil.
+func TimeFromPtr(t *time.Time) Time {
+	if t == nil {
+		var ti time.Time
+		return NewTime(ti, false)
+	}
+	return NewTime(*t, true)
+}
+
+// MarshalJSON implements json.Marshaler.
+// It will encode null if this time is null.
+func (t Time) MarshalJSON() ([]byte, error) {
+	if !t.Valid {
+		return []byte("null"), nil
+	}
+	return json.Marshal(t.Time)
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+// It supports time, string and null input.
+func (t *Time) UnmarshalJSON(data []byte) error {
+	var err error
+	var v interface{}
+	json.Unmarshal(data, &v)
+	switch x := v.(type) {
+	case time.Time:
+		t.Time = x
+	case string:
+		t.Time, err = time.Parse(time.RFC3339, x)
+	case nil:
+		t.Valid = false
+		return nil
+	default:
```

Comment:
- The difference between`json.Unmarshal` and calling `something.UnmarshalJSON` directly, is that `json.Unmarshal` will use reflection to do fancy stuff if the type doesn't have its own `UnmarshalJSON`, otherwise it calls that type's `UnmarshalJSON`. Calling it directly will probably be a few nanoseconds faster, not really a big deal.
- You're right in that checking the types for the `map[string]interface{}` case properly would be the best solution, let's return an error like good citizens.